src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,114 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] innerArray = (T[]) new Object[DEFAULT_CAPACITY];
+    private int size = 0;
+
     @Override
     public void add(T value) {
-
+        if (size == innerArray.length) {
+            grow();
+        }
+        innerArray[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index >= 0 && index <= size) {
+            if (size + 1 > innerArray.length) {
+                grow();
+            }
+            T[] arr = (T[]) new Object[size - index];
+            if (size > index) {
+                System.arraycopy(innerArray, index, arr, 0, size - index);
+            }
+            innerArray[index] = value;
+            System.arraycopy(arr, 0, innerArray, index + 1, size - index);
+            size++;
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
```

Comment:
The `addAll` method can be optimized by resizing the array once to accommodate all new elements instead of resizing it multiple times. Consider calculating the new size and calling `grow()` once before the loop.